src/mngt/ocf_mngt_common.c

```diff
@@ -175,94 +176,166 @@ int ocf_mngt_cache_get_by_id(ocf_ctx_t ocf_ctx, ocf_cache_id_t id, ocf_cache_t *
-}
+struct ocf_mngt_cache_lock_context {
+	ocf_cache_t cache;
+	ocf_unlock_fn_t unlock_fn;
+	ocf_mngt_cache_lock_end_t cmpl;
+	void *priv;
+};
 
-void ocf_mngt_cache_unlock(ocf_cache_t cache)
+static void _ocf_mngt_cache_lock_complete(
+		ocf_async_lock_waiter_t waiter, int error)
 {
-	OCF_CHECK_NULL(cache);
-	_ocf_mngt_cache_unlock(cache, env_rwsem_up_write);
+	struct ocf_mngt_cache_lock_context *context;
+	ocf_cache_t cache;
+
+	context = ocf_async_lock_waiter_get_priv(waiter);
+	cache = context->cache;
+
+	if (error) {
+		ocf_mngt_cache_put(cache);
+		goto out;
+	}
+
+	if (env_bit_test(ocf_cache_state_stopping, &cache->cache_state)) {
+		/* Cache already stopping, do not allow any operation */
+		context->unlock_fn(ocf_async_lock_waiter_get_lock(waiter));
+		ocf_mngt_cache_put(cache);
+		error = -OCF_ERR_CACHE_NOT_EXIST;
+	}
+
+out:
+	context->cmpl(context->cache, context->priv, error);
 }
 
-void ocf_mngt_cache_read_unlock(ocf_cache_t cache)
+static void _ocf_mngt_cache_lock(ocf_cache_t cache,
+		ocf_mngt_cache_lock_end_t cmpl, void *priv,
+		ocf_lock_fn_t lock_fn, ocf_unlock_fn_t unlock_fn)
 {
-	OCF_CHECK_NULL(cache);
-	_ocf_mngt_cache_unlock(cache, env_rwsem_up_read);
+	ocf_async_lock_waiter_t waiter;
+	struct ocf_mngt_cache_lock_context *context;
+
+	if (ocf_mngt_cache_get(cache))
+		OCF_CMPL_RET(cache, priv, -OCF_ERR_CACHE_NOT_EXIST);
+
+	waiter = ocf_async_lock_new_waiter(&cache->lock,
+			_ocf_mngt_cache_lock_complete);
+	if (!waiter) {
+		ocf_mngt_cache_put(cache);
+		OCF_CMPL_RET(cache, priv, -OCF_ERR_NO_MEM);
+	}
+
+	context = ocf_async_lock_waiter_get_priv(waiter);
+	context->cache = cache;
+	context->unlock_fn = unlock_fn;
+	context->cmpl = cmpl;
+	context->priv = priv;
+
+	lock_fn(waiter);
 }
 
-static int _ocf_mngt_cache_lock(ocf_cache_t cache, int (*lock_fn)(env_rwsem *s),
-		void (*unlock_fn)(env_rwsem *s))
+static int _ocf_mngt_cache_trylock(ocf_cache_t cache,
+		ocf_trylock_fn_t trylock_fn, ocf_unlock_fn_t unlock_fn)
 {
-	int ret;
+	int result;
 
-	/* Increment reference counter */
-	if (!ocf_refcnt_inc(&cache->refcnt.cache))
+	if (ocf_mngt_cache_get(cache))
 		return -OCF_ERR_CACHE_NOT_EXIST;
 
-	env_atomic_inc(&cache->lock_waiter);
-	ret = lock_fn(&cache->lock);
-	env_atomic_dec(&cache->lock_waiter);
-
-	if (ret) {
-		ocf_mngt_cache_put(cache);
-		return ret;
-	}
+	result = trylock_fn(&cache->lock);
+	if (result)
+		return result;
 
 	if (env_bit_test(ocf_cache_state_stopping, &cache->cache_state)) {
-		/* Cache already stooping, do not allow any operation */
-		ret = -OCF_ERR_CACHE_NOT_EXIST;
-		goto unlock;
+		/* Cache already stopping, do not allow any operation */
+		unlock_fn(&cache->lock);
+		return -OCF_ERR_CACHE_NOT_EXIST;
 	}
 
 	return 0;
+}
+
+static void _ocf_mngt_cache_unlock(ocf_cache_t cache,
+		ocf_unlock_fn_t unlock_fn)
+{
+	unlock_fn(&cache->lock);
+	ocf_mngt_cache_put(cache);
+}
+
+int ocf_mngt_cache_lock_init(ocf_cache_t cache)
+{
+	int result;
+
+	result = ocf_async_lock_init(&cache->lock,
```

Comment:
You can return ocf_async_lock_init() result directly instead of saving it to the temporary variable.

Reply:
Done.